sources/host-ctr/pidnamespaces/pid_namespaces.go

```diff
@@ -0,0 +1,169 @@
+		return errors.Wrap(err, "acquiring lock for writing to PID NS manifest failed with error")
+	}
+	defer func() {
+		if err = fileLock.Unlock(); err != nil {
+			err = errors.Wrap(err, "failed to release lock for writing to PID NS manifest")
+		}
+	}()
+
+	// Add task PID NS to manifest
+	manifest, err := os.OpenFile(manifestPath, os.O_RDWR|os.O_CREATE, 0664)
+	if err != nil {
+		return errors.Wrap(err, "failed to open PID NS manifest file")
+	}
+	defer manifest.Close()
+
+	manifestStat, err := manifest.Stat()
+	if err != nil {
+		return errors.Wrap(err, "failed to stat PID NS manifest file")
+	}
+	raw := make([]byte, manifestStat.Size())
+	if _, err = manifest.Read(raw); err != nil {
+		return errors.Wrap(err, "failed to read PID NS manifest file")
+	}
+	tmpPidNsList := &HostCtrPidNs{}
+	hostCtrPidNsList := &HostCtrPidNs{}
+	if len(raw) != 0 {
+		err = json.Unmarshal(raw, &tmpPidNsList)
+		if err != nil {
+			return errors.Wrap(err, "failed to unmarshal PID NS manifest")
+		}
+		for _, ele := range tmpPidNsList.PidNamespaces {
+			// If it already exists in the manifest list for whatever reason, we can just return without doing anything.
+			if ele == pidNs {
+				return nil
+			}
+		}
+	}
+	hostCtrPidNsList.PidNamespaces = append(tmpPidNsList.PidNamespaces, pidNs)
+	if err = writeManifest(hostCtrPidNsList, manifest); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RemovePidNsFromManifest takes a PID's namespace and removes it from the internal manifest
+func RemovePidNsFromManifest(pidNs int64) (err error) {
+	fileLock := flock.New(lockfile)
+	lockCtx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	ok, err := fileLock.TryLockContext(lockCtx, 500*time.Millisecond)
+	if err != nil && !ok {
+		return errors.Wrap(err, "acquiring lock for writing to PID NS manifest failed with error")
+	}
+	defer func() {
+		if err = fileLock.Unlock(); err != nil {
+			err = errors.Wrap(err, "failed to release lock for writing to PID NS manifest")
+		}
+	}()
+
+	// Remove task PID NS from manifest
+	manifest, err := os.OpenFile(manifestPath, os.O_RDWR|os.O_CREATE, 0664)
+	if err != nil {
+		return errors.Wrap(err, "failed to open PID NS manifest file")
+	}
+	defer manifest.Close()
+
+	manifestStat, err := manifest.Stat()
+	if err != nil {
+		return errors.Wrap(err, "failed to stat PID NS manifest file")
+	}
+	raw := make([]byte, manifestStat.Size())
+	if _, err = manifest.Read(raw); err != nil {
+		return errors.Wrap(err, "failed to read PID NS manifest file")
+	}
+	hostCtrPidNsList := &HostCtrPidNs{}
+	if len(raw) != 0 {
+		err = json.Unmarshal(raw, &hostCtrPidNsList)
+		if err != nil {
+			return errors.Wrap(err, "failed to unmarshal PID NS manifest")
+		}
+		for i, ele := range hostCtrPidNsList.PidNamespaces {
+			// Remove PID NS from list
+			if ele == pidNs {
+				hostCtrPidNsList.PidNamespaces = append(hostCtrPidNsList.PidNamespaces[:i], hostCtrPidNsList.PidNamespaces[i+1:]...)
+			}
+		}
```

Comment:
Is it possible to deserialize into something like a HashSet instead of an array? Then we wouldn't have to iterate across the entire list each time.

Reply:
`golang` std library is pretty bare in terms of collections. So I'd have to make my own set, something like `map[int]bool{}` with it's own interface. I was thinking the size of this would be small enough that iterating over the list of host-containers isn't too bad.

Reply:
It wouldn't need it's own interface necessarily
```go
piHashMap := make(map[int]bool)
```
or if it was the primary data structure in `hostCtrPidNs`:
```go
type HostCtrPidNs struct {
	PidNamespaces map[int64]bool `json:"pid-namespaces,omitempty"`
}
```

Then we could populate it like:
```go
// add PID to map
pidHashMap[pid] = true

// remove PID from map
delete(pidHashMap, pid)
```
instead of using append and having to use wonky slice mapping:
```go

for i, ele := range hostCtrPidNs.PidNamespaces {
	// Remove PID NS from list
	if ele == pidNs {
		hostCtrPidNs.PidNamespaces = append(hostCtrPidNs.PidNamespaces[:i], hostCtrPidNs.PidNamespaces[i+1:]...)
	}
}
```
accessing is also nice:
```go
if _, ok := pidHashMap[pid]; ok {
    // yes, pid is in map, do stuff with it here
}
```

If we did drop the map into the `hostCtrPidNs`, the tricky part would be figuring out how to deserialize it elegantly into json without abunch of `true` values with the pid keys

Reply:
There apparently is no easy way to drop map values during JSON marshaling/unmarshaling in Go without having to write my own custom marshal-er. I don't think the benefits of using a set for a relatively small size scale of host-containers is worth the additional complexity and maintenance of implementing my own marshaler. 

I'd essentially have to copy map keys to a list before marshaling and rebuild the map when unmarshaling from a list. The benefit is the `O(1)` insert vs `O(n)` insert which on the scale a few hundred host-containers (not sure people run that many host-containers) seems to matter not much. I'll put a comment to call it out as something we can revisit if this ever becomes a performance bottleneck.

Reply:
> If we did drop the map into the hostCtrPidNs, the tricky part would be figuring out how to deserialize it elegantly into json without abunch of true values with the pid keys

Hmmm thinking further on this (and chatting with Erikson), i don't know if there's a great way to marshal a struct into json as a json array without having a custom marshaler:

```go
type PidHashMap map[int64]bool

func (phm PidHashMap) MarshalJSON() ([]byte, error) {
	var out []int64
	for k := range phm {
		out = append(out, k)
	}
	return json.Marshal(out)
}
```

Reference: https://go.dev/play/p/MvPImVWHIck

Note that we would have to allocate a slice anyways in order to output only a json array without the key/values as json objects